builtin/providers/azure/resource_azure_dns_server.go

```diff
@@ -0,0 +1,246 @@
+package azure
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/Azure/azure-sdk-for-go/management/virtualnetwork"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+// resourceAzureDnsServer returns the *schema.Resource associated
+// to an Azure hosted service.
+func resourceAzureDnsServer() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAzureDnsServerCreate,
+		Read:   resourceAzureDnsServerRead,
+		Update: resourceAzureDnsServerUpdate,
+		Exists: resourceAzureDnsServerExists,
+		Delete: resourceAzureDnsServerDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Required:    true,
+				Description: parameterDescriptions["name"],
+			},
+			"dns_address": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: parameterDescriptions["dns_address"],
+			},
+		},
+	}
+}
+
+// resourceAzureDnsServerCreate does all the necessary API calls
+// to create a new DNS server definition on Azure.
+func resourceAzureDnsServerCreate(d *schema.ResourceData, meta interface{}) error {
+	// first; check for the existence of the resource:
+	exists, err := resourceAzureDnsServerExists(d, meta)
+	if err != nil {
+		return err
+	}
+	if exists {
+		return fmt.Errorf("Azure DNS server definition already exists.")
+	}
+
+	azureClient := meta.(*Client)
+	mgmtClient := azureClient.mgmtClient
+	networkClient := virtualnetwork.NewClient(mgmtClient)
+
+	log.Println("[INFO] Fetching current network configuration from Azure.")
+	azureClient.mutex.Lock()
+	defer azureClient.mutex.Unlock()
+	netConf, err := networkClient.GetVirtualNetworkConfiguration()
+	if err != nil {
+		return fmt.Errorf("Failed to get the current network configuration from Azure: %s", err)
+	}
+
+	log.Println("[DEBUG] Adding new DNS server definition to Azure.")
+	name := d.Get("name").(string)
+	address := d.Get("dns_address").(string)
+	netConf.Configuration.DNS.DNSServers = append(
+		netConf.Configuration.DNS.DNSServers,
+		virtualnetwork.DNSServer{
+			Name:      name,
+			IPAddress: address,
+		})
+
+	// send the configuration back to Azure:
+	log.Println("[INFO] Sending updated network configuration back to Azure.")
+	reqID, err := networkClient.SetVirtualNetworkConfiguration(netConf)
```

Comment:
Not a blocker, but we've tended to log a debug message with the contents of all the requests we send along so that the terraform debug logs contain field-level info on the API interaction. We've found this is very helpful in helping users debug problems.

In AWS we've been able to lean on a `StringValue` utility function that pretty prints the structs for the providers. Not sure if the Azure library has an equivalent?